### ISSUE TITLE
potential changes to the model for consideration

### ISSUE THREAD

**Issue by iantaylor-NOAA** [pacific-hake/hake-assessment#590]
Sorry for combining a bunch of things in 1 issue, but maybe this way it's easier to ignore.

I intended to spend some time in the past year exploring a bunch of issues that might be worth considering for the 2020 hake assessment but have only found time to do so very recently.

Three issues that I thought would be worth consideration by the JTC are
1. semi-parametric selectivity (recurring sensitivity in the past 2 years)
2. the use of zero-centered vs. simple deviation parameters for recruitment
3. putting a prior on the Dirichlet-Multinomial parameters to aid MCMC convergence

Here's a little bit more detail on what I've learned so far about each of these. I've uploaded some models with 24-million mcmc chains to /Hake JTC 2020/models/explorations_from_Ian/ on Google Drive. I put some comparison plots up there too.

**1. Semi-parametric selectivity**
I tried estimating the sigma_s parameter as well as fixing to 0.7 and 1.0. The estimated value was 0.77. In all cases, the recent recruitments seem implausible with the stock increasing very rapidly at the end of the time series. It would take some work to understand why the different parameterization of selectivity would be causing this, so in spite of the benefits of mcmc convergence (the diagnostics I looked at are indeed better for these semi-parametric selectivity models), I would be cautious about pursuing this option right now.

**2. zero-centered vs simple deviations for recruitment**
See discussion here: https://github.com/admb-project/admb/issues/107 about a bug in ADMB that hasn't yet been fixed. Cole is a smart guy and I think I'm finally convinced me that rather than try to fix ADMB, we should stop using zero-centered devs. Switching to "simple deviations" just requires changing line 83 of the control file to `2 #do_recdev:  0=none; 1=devvector; 2=simple deviations`.
The impact on the model is not much (B0 goes down a bit while the rest of the time series is almost the same, so relative spawning biomass goes up a bit). Ultimately, the justification for the zero-centering of the main period of recruitment devs isn't that strong and going to simple devs seems better than using results that include a known bug (even if the impact of the bug from what I can tell is even more trivial than the change in structure to simple devs).

**3. Prior on the Dirichlet-Multinomial parameter**
The transformation of the log(Theta) parameter described under `2.4.4 Age composition likelihood` in the 2019 hake assessment means that a uniform prior on log(Theta) implies a U-shaped distribution for the data weights. Inspired by Jim Thorson, I've been exploring a beta prior which would result in a flatter transformed distribution of the weights but haven't come up with the answer yet. I think the impact would be minimal on both fishery and survey weighting in the MLE context, but the prior would keep the survey weight from drifting to high values associated with weights close to 1.0 which required fixing log(Theta_survey) at the MLE to aid MCMC convergence. I can work more on this in time to explore in the assessment.

In summary, my suggestions (which you're welcome to ignore), are:
- [ ] keeping semi-parametric selectivity as a sensitivity unless you have lots of time to explore further
- [ ] switching line 83 in the control file to change to simple recruitment deviations
- [ ] wait for info about a proposed prior on log(Theta) and consider that as an alternative to fixing the survey value for MCMC

**Comment by aaronmberger-nwfsc** [comment]
Thanks Ian, this is a great summary and very useful information moving forward.  I appreciate you sharing it and continuing to stay involved in hake related research!  

Regarding #2, this will then relieve the potential 'drifting' of rec devs in mcmc's (i.e., the median rec devs from mcmc don't necessarily have to, and in fact currently never would do so exactly, sum to 0 as they do as parameter dev_vector when doing MLE), correct?  Or no, because simple deviations don't have any penalty....?

**Comment by iantaylor-NOAA** [comment]
The simple deviations (implemented as a bounded_vector rather than a bounded_dev_vector in ADMB) contain the penalty associated with sigmaR that draws all recruitment deviations toward zero in the absence of other information, but don't have the separate constraint that the "main" period has to sum to zero. 

The theory behind the use of the dev_vector with the zero sum is that the R0 and B0 reference points should be reflective of the central tendency for the time period with the best information about recruitment. This prevents the model from having "all the children are above average" as in Lake Wobegon. In theory, a mismatch between R0 and the main recdev period could lead to a better fit to the indices of abundance or some other data source which you wouldn't expect to be informative about recruitment. However, for this stock the index is pretty consistent with the recruitments and that tension doesn't seem to be present. The calculations below show that calculating the mean of the main recdevs within each of the 2000 MCMC samples gets you a distribution with median = 0.086 as compared to the status-quo approach where that would be forced to equal 0. I believe that 0.086 is a pretty insignificant difference from 0.

The issue with the ADMB sampling procedure is that the zero-sum constraint isn't actually being applied in the mcmc phase where the samples are being drawn, just the mceval phase where the results get calculated. So there's a mismatch between the samples that get reported and the sampling process that was supposed to draw them. I tried re-compiling ADMB with the proposed fix (which Cole was skeptical about) and the difference was also pretty minimal for the hake model. But even if that fix proves to be reasonable, at this stage in the 2020 hake assessment, an unproven modification to ADMB seems like a less stable path forward than the alternative approach of using simple recruit devs instead of the zero-sum dev_vector.

```
# read 2019 model modifieid to have simple devs instead of dev_vector
model <- SS_output(file.path(mydir, 'hake2019_3.30.14.08_simple_devs'),
                   dir.mcmc = 'mcmc')

# get matrix of main recdevs (1970-2017) for each of the 2000 MCMC draws
main_recdev_matrix_MCMC <- model$mcmc[ , grep("Main_RecrDev", names(model$mcmc))]
# get MLE values of main recdevs
main_recdev_vector_MLE <- model$parameters[grep("Main_RecrDev",
                                                rownames(model$parameters)),
                                           "Value"]
# get mean of main recdevs for each of the MCMC draws
main_recdev_means_MCMC <- apply(main_recdev_matrix_MCMC, MARGIN = 1, FUN = mean)

# summary statistics
mean(main_recdev_vector_MLE)
## [1] 0.213518
summary(main_recdev_means_MCMC)
##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
## -0.50835 -0.02647  0.08643  0.09200  0.21289  0.72256
```

**Comment by kellijohnson-NOAA** [comment]
Thanks @iantaylor-NOAA. I moved your exploration relating to (2) into the main model folder for 2020 and labeled it 2019.03.92_simpledevs. I hope that is okay. I also added the model to the spreadsheet of 2020 models. 

Below is some text that I drafted for the assessment regarding the change. I think it might be too long winded. If others can please edit it for me, that would be great. Just copy and paste it into a comment and edit away.

The parameterization of recruitment deviations was changed from using a zero-centered vector of deviations during the main period to a simple vector that allowed the estimates from the main period to have a non-zero sum. This change to a simple vector was necessary because a mismatch was found between the mean of the estimates from the MCMC sampling algorithm and the MCMC evaluation algorithm in ADMB. A simple vector avoids this mismatch caused by ADMB and allows the penalty towards zero to operate on the entire vector of estimated recruitment deviations rather than assuming the main period should sum to zero. Penalties have implications for calculating derivatives, and thus, we believe the use of a simple vector is currently the best way forward. Additionally, removal of the mismatch in the MCMC protocol should, in theory, lead to increased similarity between the MLE and MCMC results.

**Comment by aaronmberger-nwfsc** [comment]
I think we can close this for now, even though it has helpful explanations in here, we can always search it back out.
